layabout.py

```diff
@@ -1,3 +1,287 @@
+
+        Args:
+            type: The name of a Slack RTM API event to be handled. As a
+                special case, although it is not a proper RTM event, ``*`` may
+                be provided to handle all events. For more information about
+                available events see the
+                `Slack RTM API <https://api.slack.com/rtm>`_.
+            kwargs: Optional arbitrary keyword arguments passed to the event
+                handler when the event is triggered.
+
+        Returns:
+            A decorator that validates and registers a Layabout event handler.
+
+        Raises:
+            TypeError: If the decorated :obj:`Callable`'s signature does not
+                accept at least 2 parameters.
+        """
+        def decorator(fn: Callable) -> Callable:
+            # Validate that the wrapped callable is a suitable event handler.
+            sig = signature(fn)
+            if len(sig.parameters) < 2:
+                raise TypeError("Layabout event handlers take at least 2 "
+                                "positional arguments, a slack client and an "
+                                "event")
+
+            # Register a tuple of the callable and its kwargs, if any.
+            self._handlers[type].append((fn, kwargs or {}))
+
+            @wraps(fn)
+            def wrapper(slack: SlackClient, event: Dict[str, Any],
+                        **kwargs: Any) -> Any:
+                # We don't actually do anything with the return value, but this
+                # might make unit testing easier for users.
+                return fn(slack, event, **kwargs)
+            return wrapper
+        return decorator
+
+    def connect(self, token: str = None) -> bool:
+        """
+        Attempt to establish or reset a connection to Slack's API.
+
+        Note:
+            It isn't normally necessary to use this method as
+            :obj:`Layabout.run` will take care of establishing a connection for
+            you. This is mostly helpful if you want to take advantage of the
+            embedded :obj:`slackclient.SlackClient` on the :obj:`Layabout`
+            instance *before* entering into the :obj:`Layabout.run` loop.
+
+        Args:
+            token: A Slack API token. If given it will override an existing
+                connection (if one exists), otherwise an existing token or
+                an environment variable will be used.
+
+        Returns:
+            Whether the connection succeeded.
+
+        Raises:
+            MissingSlackToken: If no API token is available.
+        """
+        resetting = False
+
+        # If we were given a token let's prefer the new one and establish or
+        # reset the connection.
+        if token is not None:
+            self._token = token
+            resetting = True
+
+        # We don't have an existing token, so let's try to get one from an
+        # environment variable or give up.
+        if self._token is None:
```

Comment:
Ah. I think what you're talking about would need to be accomplished through a [`@property`](https://docs.python.org/3.6/library/functions.html#property) in Python.
```python
class Layabout:
    def __init__(self, env_var='SLACK_API_TOKEN'):
        self._token = None
        self._env_var

    @property
    def token(self):
        return self._token or os.getenv(self._env_var)

    @token.setter
    def token(self, value):
        self._token = value
```

This way you could do
```python
>>> layabout = Layabout()
>>> layabout.token  # 'SLACK_API_TOKEN' is undefined.
None
>>> layabout.token = 'this is not a slack token'
>>> layabout.token
'this is not a slack token'
>>> layabout.token = None  # Reset the token.
>>> os.environ['SLACK_API_TOKEN'] = 'this might be a slack token'
>>> layabout.token
'this might be a slack token'
```
This is powerful, but I also generally avoid it because I believe it violates the [Zen of Python](https://www.python.org/dev/peps/pep-0020/#id3) mantra of
> Simple is better than complex.

and it can be especially surprising when what you thought was a simple attribute lookup ends up executing a whole mess of stuff.

However, in this case, if I set the inner `token` as `self.__token` and create a `@property` of `_token`, then it's still "private" and can be left as an implementation detail giving me all of the power and none of the responsibility (muahaha!). Thoughts, @RadicalZephyr?

Reply:
If you have the time, I'm curious about your thoughts here as well, @rawrgulmuffins.

Reply:
I don't think it needs to be a property. Seems to me, you could leave the default token value as `None`, then do something like 
```
if token:
    self.slack = None
self._token = token or self.get_token()
```

IMO, I think the main reason to introduce a property is when you had an attribute as part of your public API, and you want to change it to be a function call without breaking code.